onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -241,6 +302,77 @@ Status ReduceKernel<allow_multi_axes>::ComputeImpl(OpKernelContext* ctx, cudnnRe
   return Status::OK();
 }
 
+template <>
+template <>
+Status ReduceKernel<true>::ComputeImpl<int32_t, CUDNN_REDUCE_TENSOR_NO_INDICES>(OpKernelContext* ctx, cudnnReduceTensorOp_t cudnnReduceOp) const {
+  typedef typename ToCudaType<int32_t>::MappedType CudaT;
+
+  const Tensor* X = nullptr;
+  Tensor* Y = nullptr;
+
+  int64_t input_count = 0;
+  int64_t output_count = 0;
+  std::vector<int64_t> input_dims_cudnn;
+  std::vector<int64_t> output_dims_cudnn;
+  std::vector<int64_t> output_dims_keepdims;
+  ORT_RETURN_IF_ERROR(PrepareForReduce(ctx,
+                                       keepdims_,
+                                       axes_,
+                                       &X,
+                                       &Y,
+                                       input_count,
+                                       output_count,
+                                       input_dims_cudnn,
+                                       output_dims_cudnn,
+                                       output_dims_keepdims));
+
+  // cudnnReduceTensor for ReduceSum has issue if input and output has same size, we just need to copy the data for this case
+  if (input_count == output_count) {
+    if (Y->template MutableData<int32_t>() != X->template Data<int32_t>()) {
+      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(Y->template MutableData<int32_t>(), X->template Data<int32_t>(), input_count * sizeof(int32_t), cudaMemcpyDeviceToDevice));
+    }
+    return Status::OK();
+  }
+
+  size_t indices_bytes = 0;
+  size_t workspace_bytes = 0;
+  CudnnTensor input_tensor;
+  CudnnTensor output_tensor;
+  CudnnReduceDescriptor reduce_desc;
+
+  cudnnDataType_t cudnn_type_X = CUDNN_DATA_FLOAT;
+  IAllocatorUniquePtr<float> temp_X = GetScratchBuffer<float>(input_count);
+  Impl_Cast<CudaT, float>(reinterpret_cast<const CudaT*>(X->template Data<int32_t>()), temp_X.get(), X->Shape().Size());
```

Comment:
Could you add some comments about why the cast is needed? Plus, int32->float->int32 is lossy and may have accuracy issues for int32 > 2^23